File: Java/HelloJ/MySort.java
import java.util.Comparator;
import java.util.Stack;

public class MySort {
    public<T extends  Comparable<T>> void sort(T[] list){
        boolean swapped = true;
        for (int i = 1,len = list.length;i<len&&swapped;i++){
            swapped = false;
            for(int j = 0;j<len-1;j++){
                if(list[j].compareTo(list[j+1])>0){
                    T temp = list[j];
                    list[j]=list[j+1];
                    list[j+1]=temp;
                    swapped=true;
                }
            }
        }
    }
    public <T> void sort(T[] list,Comparator<T> comp){
        boolean swapped = true;
        for (int i = 1,len = list.length;i<len&&swapped;i++){
            swapped = false;
            for(int j = 0;j<len-1;j++){
                if(comp.compare(list[j],list[j+1])>0){
                    T temp = list[j];
                    list[j]=list[j+1];
                    list[j+1]=temp;
                    swapped=true;
                }
            }
        }
    }

    //直接插入排序
    public static void insertiontSort(int[] arr){
        for(int i = 1;i<arr.length;i++){
            int temp = arr[i];
            for(int j = i;j>0;j--){
                if(arr[j-1]>temp)
                    arr[j]=arr[j-1];
                else
                    arr[j]=temp;
            }
        }
    }

    //希尔排序
    public static void shellSort(int[] arr){
        int gap =1;
        int temp;
        while (gap<arr.length/3)
            gap= gap*3+1;
        while (gap>=1){
            for(int i = gap;i<arr.length;i++){
                for(int j = i;j>=gap&&arr[j]<arr[j-gap];j-=gap){
                    temp = arr[j];
                    arr[j]=arr[j-gap];
                    arr[j-gap]=temp;
                }
            }
            gap = gap/3;
        }

    }

    //选择排序
    public static void selectionSort(int[] arr){
        int min;
        for(int i = 0;i<arr.length-1;i++){
           min = i;
            for(int j =i;j<arr.length;j++){
                if(arr[j]<arr[min])
                    min = j;
            }
            int temp = arr[min];
            arr[min]=arr[i];
            arr[i]=temp;
        }
    }

    //堆排序
    private static void heapAjd(int[] arr,int m,int len){
        int left = 2*m +1;
        int right =2*m +2;
        int large = m;
        if(left<len&&arr[left]>arr[large])
            large = left;
        if(right<len&&arr[right]>arr[large])
            large=right;
        if(large!=m){
            swap(arr,m,large);
            heapAjd(arr,large,len);
        }
    }
    public static void heapSort(int[] arr){
        int i,temp;
        int len = arr.length-1;
        for(i = arr.length/2;i>=0;i--)
            heapAjd(arr,i,len);
        for(i = arr.length-1;i>=0;i--){
            swap(arr,0,i);
            heapAjd(arr,0,len--);
        }
    }

    private static void swap(int[] arr,int a,int b){
        int temp = arr[a];
        arr[a]=arr[b];
        arr[b]=temp;
    }

    //冒泡排序
    public static void bubbbleSort(int[] arr){
        for(int i = 0;i<arr.length-1;i++){
            for(int j = 0;j<arr.length-i-1;j++){
                if(arr[j]>arr[j+1])
                    swap(arr,j,j+1);
            }
        }
    }

    //快速排序
    public static void quickSort(int arr[]){
        qSort(arr,0,arr.length-1);
    }
    private static void qSort(int arr[],int left,int right){
        int pivot;
        if(left<right){
            pivot= partition(arr,left,right);
            qSort(arr,left,pivot-1);
            qSort(arr,pivot+1,right);
        }
    }
    private static int partition(int arr[],int left,int right){
        int key = arr[left];
        while(left<right){
            while(left<right&&arr[right]>=key)
                right--;
            swap(arr,left,right);
            while(left<right&&arr[left]<=key)
                left++;
            swap(arr,left,right);
        }
        return left;
    }
    //使用栈代替递归
    public static void quickSortByStack(int[] arr){
        if(arr.length <= 0) return;
        Stack<Integer> stack = new Stack<Integer>();

        stack.push(0);
        stack.push(arr.length-1);
        int low,high,pivot;
        while(!stack.isEmpty()){
            high = stack.pop();
            low = stack.pop();
            pivot = partition(arr,low,high);
            if(pivot>low){
                stack.push(low);
                stack.push(pivot-1);
            }
            if(pivot<high){
                stack.push(pivot+1);
                stack.push(high);
            }
        }
    }

    //归并排序
    public static void mergeSort(int arr[]){
        mSort(arr,0,arr.length-1);
    }
    private static void mSort(int arr[],int left,int right){
        if(left==right) return;
        int m = (left+right)/2;
        mSort(arr,left,m);
        mSort(arr,m+1,right);
        merge(arr,left,m,right);
    }
    private static void merge(int arr[],int left,int m,int right){
        int[] tr=new int[right-left+1];
        int i,j,k;
        for(k=0,i = left,j = m+1;i<=m&&j<=right;k++){
            if(arr[i]<=arr[j]){
                tr[k]=arr[i];
                i++;
            }
            else{
                tr[k]=arr[j];
                j++;
            }
        }
        while(i!=m+1){
            tr[k]=arr[i];
            i++;
            k++;
        }
        while(j!=right+1){
            tr[k]=arr[j];
            j++;
            k++;
        }
        for(i=0;i<=right-left;i++){
            arr[left+i]=tr[i];
        }
    }
    //基数排序
    //lsd
    public static void radixSort(int arr[]){
        int max = -1;
        for(int i = 0;i<arr.length;i++){
            if(max<arr[i])
                max = arr[i];
        }
        int maxDigit=1;
        while(max/10>0){
            maxDigit++;
            max/=10;
        }
        int[][] buckets = new int[10][arr.length];
        int base = 10;

        for(int i = 0;i<maxDigit;i++){
            int[] bktLen = new int[10];
            //分配
            for(int j = 0;j<arr.length;j++){
                int whickBucket = (arr[j]%base)/(base/10);
                buckets[whickBucket][bktLen[whickBucket]]=arr[j];
                bktLen[whickBucket]++;
            }
            //收集
            int k = 0;
            for(int b = 0;b<buckets.length;b++){
                for(int p=0;p<bktLen[b];p++){
                    arr[k++]=buckets[b][p];
                }
            }
            base*=10;
        }
    }
}

File: C++/HelloCPP/mySort.cpp
#include "mySort.h"
#include "util.cpp"

//下一步
//将全部转化为模板方法

void bubbleSort(int *l, int n)
{
    for (int i = 0; i < n - 1; i++)
    {
        for (int j = 0; j < n - 1; j++)
        {
            if (l[j] > l[j + 1])
                mySwap(l[j + 1], l[j]);
        }
    }
}
void sampleSelectSort(int *l, int n)
{
    int k;
    for (int i = 0; i < n - 1; i++)
    {
        k = i;
        for (int j = i + 1; j < n; j++)
        {
            if (l[j] < l[k])
                k = j;
        }
        mySwap(l[i], l[k]);
    }
}
void sampleInsertSort(int *l, int n)
{

    for (int i = 1; i < n; i++)
    {
        for (int j = i - 1; (j >= 0) && (l[i] < l[j]); j--)
        {
            mySwap(l[i], l[j]);
        }
    }
}

void shellSort(int *l, int n)
{
    int incre = 1;
    while (incre < n / 3)
    {
        incre = incre * 3 + 1;
    }
    while (incre)
    {
        for (int i = incre; i < n; i += incre)
        {
            for (int j = i - incre; (j >= 0) && (l[i] < l[j]); j -= incre)
            {
                mySwap(l[i], l[j]);
            }
        }
        incre /= 3;
    }
}

void merge(int *l, int left, int mid, int right)
{
    int st[right - left];
    int i = left, j = mid, k = 0;
    for (i, j, k; (i < mid) && (j < right); k++)
    {
        if (l[i] <= l[j])
        {
            st[k] = l[i];
            i++;
        }
        else
        {
            st[k] = l[j];
            j++;
        }
    }
    while (i < mid)
    {
        st[k++] = l[i++];
    }
    while (j < right)
    {
        st[k++] = l[j++];
    }
    for (k = left; k < right - left; k++)
        l[k] = st[k];
}
void Msort(int *l, int left, int right)
{
    if (right - left <= 1)
        return;
    int mid = (left + right) / 2;
    Msort(l, left, mid);
    Msort(l, mid, right);
    merge(l, left, mid, right);
}
void mergeSort(int *l, int n)
{
    Msort(l, 0, n);
}

int partition(int *l, int left, int right)
{
    int key = l[left];
    while (right - left > 1)
    {
        while ((right - left > 1) && (l[right - 1]) >= key)
            right--;
        mySwap(l[left], l[right - 1]);
        while ((right - left > 1) && (l[left] < key))
            left++;
        mySwap(l[right - 1], l[left]);
    }
    return left;
}
void Qsort(int *l, int left, int right)
{
    if (right - left > 1)
    {
        int pivot = partition(l, left, right);
        Qsort(l, left, pivot);
        Qsort(l, pivot + 1, right);
    }
}
void quickSort(int *l, int n)
{
    Qsort(l, 0, n);
}

//将 m 到 n 构造为大顶堆
void heapAdjust(int *l, int m, int n)
{
    int left = 2 * m + 1;
    int right = 2 * m + 2;
    int large = m;
    if (left < n && l[left] > l[large])
        large = left;
    if (right < n && l[right] > l[large])
        large = right;
    if (large != m)
    {
        mySwap(l[m], l[large]);
        heapAdjust(l, large, n);
    }
}
void heapAdj(int *l, int m, int n)
{
    int large = m;
    for (int i = 2 * m; i < n; i *= 2)
    {
        if (i < n && l[i] < l[i + 1])
            i++;
        if (l[i] <= l[large])
            break;
        mySwap(l[large], l[i]);
        large = i;
    }
}
void heapSort(int *l, int n)
{
    //构造大顶堆
    int i;
    for (i = n / 2; i >= 0; i--)
    {
        heapAdj(l, i, n);
    }
    //将最上面的取出，重新构造小顶堆
    for (i = n - 1; i > 0; i--)
    {
        mySwap(l[0], l[i]);
        heapAdj(l, 0, i - 1);
    }
}

//test
// #include <iostream>
// using namespace std;
// void printL(int *, int);
// int main()
// {
//     int L[] = {4, 2, 4, 6, 2, 6, 7, 20, 9, 12};
//     int len = sizeof(L) / sizeof(L[0]);
//     heapSort(L, len);
//     printL(L, len);
//     return 0;
// }
// void printL(int *L, int len)
// {
//     for (int i = 0; i < len; i++)
//         cout << L[i] << ",";
//     cout << endl;
// }